config.txt
AUTOTRADE=True
TICKERS=''
THRESHOLD=5
MIN_VOLUME=10000000
OPTION_MOVING_AVERAGE=15m
MAX_AUTO_TRADES=2
PROFIT_PERCENTAGE=10
LOSS_PERCENTAGE=50
PROFIT_AMOUNT=0.00
LOSS_AMOUNT=0.00
LEVERAGE=20
ORDER_AMOUNT_PERCENTAGE=10
BARS=100
MA_FAST=8
MA_MEDIUM=13
MA_SLOW=21
RSI_FAST=6
RSI_SLOW=14
BBM_PERIOD=20
BBM_STD=2
MACD_PERIOD=9
MACD_SHORT=12
MACD_LONG=26
ADX_PERIOD=21
OPEN_ON_ANY_SIGNAL=True
EMA_STUDY=True
EMA_CHART=True
EMA_CROSSING=False
EMA_CHECK_ON_OPEN=True
EMA_CHECK_ON_CLOSE=False
EMA_CLOSE_ON_FAST_MEDIUM=True
MACD_STUDY=True
MACD_CHART=True
MACD_CROSSING=True
MACD_CHECK_ON_OPEN=False
MACD_CHECK_ON_CLOSE=False
BBM_STUDY=True
BBM_CHART=False
BBM_CROSSING=False
BBM_CHECK_ON_OPEN=False
BBM_CHECK_ON_CLOSE=False
RSI_STUDY=True
RSI_CHART=True
RSI_CROSSING=False
RSI_CHECK_ON_OPEN=False
RSI_CHECK_ON_CLOSE=False
BOS_PERIOD=24
BOS_STUDY=True
BOS_CHART=True
BOS_CHECK_ON_OPEN=True
BOS_CHECK_ON_CLOSE=False
TRENDLINE_PEAK_DISTANCE=4
TRENDLINE_STUDY=True
TRENDLINE_CHART=True
TRENDLINE_BREAKOUT=True
TRENDLINE_CHECK_ON_OPEN=False
TRENDLINE_CHECK_ON_CLOSE=False
ADX_STUDY=True
ADX_CHECK_ON_OPEN=True
ADX_CHECK_ON_CLOSE=False
CANDLE_TREND_STUDY=True
CANDLE_TREND_CHECK_ON_OPEN=False
CANDLE_TREND_REVERSAL_CHECK_ON_CLOSE=True
SCREEN_REFRESH_INTERVAL=1
SIGNAL_CHECK_INTERVAL=15
PORTFOLIO_API_INTERVAL=1
PENDING_POSITIONS_API_INTERVAL=1
PENDING_ORDERS_API_INTERVAL=3
TRADE_HISTORY_API_INTERVAL=3
POSITION_HISTORY_API_INTERVAL=3
TICKER_DATA_API_INTERVAL=15
BATCH_PROCESS_SIZE=2500
USE_PUBLIC_WEBSOCKET=True
VERBOSE_LOGGING=False
CPU_PROCCESING=True
BENCHMARK=False


bitunix_automated_crypto_trading/config.txt
AUTOTRADE=True
TICKERS=''
THRESHOLD=100
MIN_VOLUME=100000000
OPTION_MOVING_AVERAGE=1h
MAX_AUTO_TRADES=5
PROFIT_PERCENTAGE=2
LOSS_PERCENTAGE=15
PROFIT_AMOUNT=0.00
LOSS_AMOUNT=0.00
BOT_CONTROLS_TP_SL=False
BOT_TRAIL_TP=True
BOT_TRAIL_SL=True
TP_STOP_TYPE=MARK_PRICE
TP_ORDER_TYPE=LIMIT
SL_STOP_TYPE=MARK_PRICE
SL_ORDER_TYPE=MARKET
LEVERAGE=20
ORDER_AMOUNT_PERCENTAGE=10
DELAY_IN_MINUTES_FOR_SAME_TICKER_TRADES=5
BARS=100
MA_FAST=8
MA_MEDIUM=13
MA_SLOW=21
RSI_FAST=6
RSI_SLOW=14
BBM_PERIOD=20
BBM_STD=2
MACD_PERIOD=9
MACD_SHORT=12
MACD_LONG=26
ADX_PERIOD=21
MINIMUM_CONSECUTIVE_CANDLES=2
OPEN_ON_ANY_SIGNAL=False
EMA_STUDY=True
EMA_CHART=True
EMA_CROSSING=False
EMA_CHECK_ON_OPEN=True
EMA_CHECK_ON_CLOSE=False
EMA_CLOSE_ON_FAST_MEDIUM=True
MACD_STUDY=True
MACD_CHART=True
MACD_CROSSING=False
MACD_CHECK_ON_OPEN=False
MACD_CHECK_ON_CLOSE=False
BBM_STUDY=True
BBM_CHART=True
BBM_CROSSING=False
BBM_CHECK_ON_OPEN=False
BBM_CHECK_ON_CLOSE=False
RSI_STUDY=True
RSI_CHART=True
RSI_CROSSING=False
RSI_CHECK_ON_OPEN=False
RSI_CHECK_ON_CLOSE=False
BOS_PERIOD=24
BOS_STUDY=True
BOS_CHART=True
BOS_CHECK_ON_OPEN=False
BOS_CHECK_ON_CLOSE=False
TRENDLINE_PEAK_DISTANCE=24
TRENDLINE_STUDY=True
TRENDLINE_CHART=True
TRENDLINE_BREAKOUT=True
TRENDLINE_CHECK_ON_OPEN=False
TRENDLINE_CHECK_ON_CLOSE=False
ADX_STUDY=True
ADX_CHECK_ON_OPEN=True
ADX_CHECK_ON_CLOSE=False
CANDLE_TREND_STUDY=True
CANDLE_TREND_CHECK_ON_OPEN=True
CANDLE_TREND_REVERSAL_CHECK_ON_CLOSE=False
SCREEN_REFRESH_INTERVAL=1
SIGNAL_CHECK_INTERVAL=15
PORTFOLIO_API_INTERVAL=1
PENDING_POSITIONS_API_INTERVAL=1
PENDING_ORDERS_API_INTERVAL=1
TRADE_HISTORY_API_INTERVAL=3
POSITION_HISTORY_API_INTERVAL=3
TICKER_DATA_API_INTERVAL=15
BATCH_PROCESS_SIZE=2500
USE_PUBLIC_WEBSOCKET=True
VERBOSE_LOGGING=False
CPU_PROCCESING=True
BENCHMARK=False


bitunix_automated_crypto_trading/config.py
from pydantic import Field, SecretStr, validator
from pydantic_settings import BaseSettings
import os
import sys
import sqlite3

class Settings(BaseSettings):
    # Start autotrading on start
    AUTOTRADE: bool = Field(default=False)
    
    #Ticker list
    TICKERS:str = Field(default="")
    THRESHOLD: float = Field(default=100.0, ge=0.0)
    MIN_VOLUME: int = Field(default=500000000, ge=0)

    # Trading Parameters
    LEVERAGE: int = Field(default=20, ge=1, le=100)
    ORDER_AMOUNT_PERCENTAGE: float = Field(default=10, ge=0.0, le=100)
    MAX_AUTO_TRADES: int = Field(default=10, ge=0)
    PROFIT_PERCENTAGE: float = Field(default=10, ge=0.0)
    LOSS_PERCENTAGE: float = Field(default=10, ge=0.0)
    PROFIT_AMOUNT: float = Field(default=0, ge=0.0)
    LOSS_AMOUNT: float = Field(default=5.0, ge=0.0)
    BOT_CONTROLS_TP_SL: bool = Field(default=False)
    BOT_TRAIL_TP: bool = Field(default=False)
    BOT_TRAIL_SL: bool = Field(default=False)
    TP_STOP_TYPE: str = Field(default="MARK_PRICE")
    TP_ORDER_TYPE: str = Field(default="LIMIT")
    SL_STOP_TYPE: str = Field(default="MARK_PRICE")
    SL_ORDER_TYPE: str = Field(default="MARKET")
    OPTION_MOVING_AVERAGE: str = Field(default="1h")
    DELAY_IN_MINUTES_FOR_SAME_TICKER_TRADES: int = Field(default=15, ge=0)
    BARS: int = Field(default=100, ge=1)

    # Technical Indicators Parameters
    MA_FAST: int = Field(default=10, ge=1)
    MA_MEDIUM: int = Field(default=20, ge=1)
    MA_SLOW: int = Field(default=50, ge=1)
    RSI_FAST: int = Field(default=6, ge=1)
    RSI_SLOW: int = Field(default=24, ge=1)
    BBM_PERIOD: int = Field(default=20, ge=1)
    BBM_STD: int = Field(default=2, ge=1)
    MACD_PERIOD: int = Field(default=9, ge=1)
    MACD_SHORT: int = Field(default=12, ge=1)
    MACD_LONG: int = Field(default=26, ge=1)
    ADX_PERIOD: int = Field(default=14, ge=1)
    MINIMUM_CONSECUTIVE_CANDLES: int = Field(default=2, ge=1)

    # Technical Indicators
    OPEN_ON_ANY_SIGNAL: bool = Field(default=True)

    EMA_CHART: bool = Field(default=True)
    EMA_STUDY: bool = Field(default=True)
    EMA_CROSSING: bool = Field(default=False)
    EMA_CHECK_ON_OPEN: bool = Field(default=True)
    EMA_CHECK_ON_CLOSE: bool = Field(default=True)
    EMA_CLOSE_ON_FAST_MEDIUM: bool = Field(default=False)

    MACD_CHART: bool = Field(default=True)
    MACD_STUDY: bool = Field(default=True)
    MACD_CROSSING: bool = Field(default=True)
    MACD_CHECK_ON_OPEN: bool = Field(default=True)
    MACD_CHECK_ON_CLOSE: bool = Field(default=False)

    BBM_CHART: bool = Field(default=True)
    BBM_STUDY: bool = Field(default=True)
    BBM_CROSSING: bool = Field(default=True)
    BBM_CHECK_ON_OPEN: bool = Field(default=True)
    BBM_CHECK_ON_CLOSE: bool = Field(default=False)

    RSI_CHART: bool = Field(default=True)
    RSI_STUDY: bool = Field(default=True)
    RSI_CROSSING: bool = Field(default=True)
    RSI_CHECK_ON_OPEN: bool = Field(default=True)
    RSI_CHECK_ON_CLOSE: bool = Field(default=False)

    BOS_PERIOD: int = Field(default=24, ge=1)
    BOS_STUDY: bool = Field(default=True)
    BOS_CHART: bool = Field(default=True)
    BOS_CHECK_ON_OPEN: bool = Field(default=True)
    BOS_CHECK_ON_CLOSE: bool = Field(default=False)

    TRENDLINE_PEAK_DISTANCE: int = Field(default=3, ge=0, le=30)
    TRENDLINE_CHART: bool = Field(default=True)
    TRENDLINE_STUDY: bool = Field(default=True)
    TRENDLINE_BREAKOUT: bool = Field(default=True)
    TRENDLINE_CHECK_ON_OPEN: bool = Field(default=True)
    TRENDLINE_CHECK_ON_CLOSE: bool = Field(default=False)

    ADX_STUDY: bool = Field(default=True)
    ADX_CHECK_ON_OPEN: bool = Field(default=True)
    ADX_CHECK_ON_CLOSE: bool = Field(default=False)

    CANDLE_TREND_STUDY: bool = Field(default=True)
    CANDLE_TREND_CHECK_ON_OPEN: bool = Field(default=True)
    CANDLE_TREND_REVERSAL_CHECK_ON_CLOSE: bool = Field(default=False)

    # Time Intervals
    SCREEN_REFRESH_INTERVAL: int = Field(default=1, ge=1)
    SIGNAL_CHECK_INTERVAL: int = Field(default=15, ge=1)
    PORTFOLIO_API_INTERVAL: int = Field(default=3, ge=1)
    PENDING_POSITIONS_API_INTERVAL: int = Field(default=3, ge=1)
    PENDING_ORDERS_API_INTERVAL: int = Field(default=3, ge=1)
    TRADE_HISTORY_API_INTERVAL: int = Field(default=3, ge=1)
    POSITION_HISTORY_API_INTERVAL: int = Field(default=3, ge=1)
    TICKER_DATA_API_INTERVAL: int = Field(default=30, ge=1)
    BATCH_PROCESS_SIZE: int = Field(default=1000, ge=1)

    # Use websocket or API
    USE_PUBLIC_WEBSOCKET: bool = Field(default=False)  # If there is lagging issue then use API

    # Logger
    VERBOSE_LOGGING: bool = Field(default=False)

    #use CPU
    CPU_PROCCESING: bool = Field(default=True)

    # Benchmark
    BENCHMARK: bool = Field(default=False)
    
    #DB
    DATABASE: str = Field(default="bitunix.db")
    

    class Config:
        # Specify the file name for loading environment variables
        config_file=sys.argv[2] if len(sys.argv) > 1 else "config.txt"
        env_file = os.path.dirname(os.path.abspath(__file__))+"/"+config_file

   
    @classmethod
    def load_from_db(cls, db_path: str, table_name: str = "settings"):
        # Connect to SQLite database
        connection = sqlite3.connect(db_path)
        cursor = connection.cursor()

        # Fetch all parameters and values from the database
        cursor.execute(f"SELECT param, value FROM {table_name}")
        rows = cursor.fetchall()
        connection.close()

        # Map fetched data to the class attributes dynamically
        settings_dict = {}
        for param, value in rows:
            # Convert string 'value' to proper type based on attribute type
            if hasattr(cls, param):
                attr_type = type(getattr(cls(), param))
                settings_dict[param] = attr_type(value) if attr_type != bool else value.lower() in ("true", "1")

        # Create the Settings instance with mapped data
        return cls(**settings_dict)

